[Alvin Tan Fu Long] iP

alvintfl opened the pull request:
# Tako
> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes/))

Tako frees your mind of having to remember things you need to do. It's,
* text-based
* easy to learn
* ~~FAST~~ *SUPER* FAST to use

All you need to do is,
1. download it from [here.](https://github.com/alvintfl/ip/releases)
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉 

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing dealines (coming soon)
- [ ] Reminders (coming soon)
-----------------------------------------
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

alvintfl committed:
Tako class: change main method to echo user commands

Add feature following Level-1 instructions.
Rename Duke to Tako for unique name.

alvintfl committed:
Rename EXIT_PROGRAM to EXIT_COMMAND

alvintfl committed:
Tako class: add list command and store user input

Add feature following level-2 instructions.

alvintfl committed:
Add Task class

Package states and functionality related to a task.

alvintfl committed:
Tako class: add task list functionality in main method

Add feature following level-3 instructions.

alvintfl committed:
Add Todo, Deadline and Event classes

There are multiple type of tasks with similarities between them.

Let's use inheritance to reduce code duplication and make use of
polymorphism.

alvintfl committed:
Fix previous commit where files were not staged

alvintfl committed:
Tako class: Add support for tracking ToDos, Deadlines & Events

There are different type of tasks to track.

Let's add new commands to track the different tasks.

alvintfl committed:
Tako class: Do automated text UI testing

Ensure that Tako class works as intended.

Let's test all valid commands.

alvintfl committed:
Add Tako specific exceptions

To handle errors caused by users using Tako incorrectly.

Let's add exceptions to deal with the following:
* empty description for commands related to creating a task
* invalid input
* marking a task that is out of range

alvintfl committed:
Tako class: throw and handle exceptions

Tako stops working when an invalid input is given.

Let's allow Tako to throw exceptions when an invalid input is given
and then handle it to allow Tako to continue working.

alvintfl committed:
Tako class: use Arraylist instead of an array to store tasks

Arraylist can be used to store data instead of just an array.

alvintfl committed:
Tako class: add support for deleting tasks from list

Add feature following level-6 instructions.

alvintfl committed:
Add Command enum

Since there are only a fixed set of commands, we can use an enum to
represent the commands.

Let's change all command constants to values under the Command
enum. User input is now trimmed and a static contains method is
added to Command enum.

This allows us to convert user input from type String to Command.

alvintfl committed:
Bye command: only works if "bye" is inputted.

To be consistent with how the list command works.

alvintfl committed:
Tako class: Add save and load methods

The tasks tracked by Tako are not saved when the program is closed,
causing Tako to lose all tracked tasks.

Let's,
* add a method to save the tasks on a file in the user's computer
* add a method to load the same tasks on starting Tako

alvintfl committed:
Use LocalDateTime to handle dates and times.

Dates and times are represented using strings which lack the
features of LocalDateTime that Tako can use.

Let's change Deadline and Event to store dates and times as
LocalDateTime. Tako should also be changed to take in date and time
inputs and parse them as LocalDateTime.

alvintfl committed:
Tako class: Add saveToFile() method

The save() method rewrites the whole file even though only one task
is added.

This is inefficient when there are multiple tasks in the task list.

Let's,
* add a saveToFile() method to append to the file instead
* abstract out the file formatting details from save() such that
saveToFile() and saveTasksToFile() methods can both call a method
to do the file formatting

alvintfl committed:
Merge branch 'branch-Level-7'

alvintfl committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Tako.java

alvintfl committed:
Refactor Tako class into different classes

It is hard to maintain Tako as it does everything by itself.

Let's,
* refactor Tako into different classes with only one
responsibility
* refactor previous exceptions into one TakoException since the way
exceptions are handled is the same
* add abstract command class so that each command can handle their
own behaviour

alvintfl committed:
Put all classes into a java package

There is a lack of organisation where all the classes are in the
same folder.

Let's put all the classes into a java package.

alvintfl committed:
Merge remote-tracking branch 'origin/add-gradle-support'

alvintfl committed:
Set up Gradle and checkstyle on Gradle

Automate build tasks of Tako and make sure code follows style
guide.

alvintfl committed:
Add unit tests for Parser and TaskList

Ensure that public methods for Parser and TaskList work as
expected.

alvintfl committed:
TaskListTest: add space after throws TakoException

alvintfl committed:
Add JavaDoc for all classes and public non-trivial methods.

alvintfl committed:
Add FindCommand to find tasks.

Allow the user to find tasks related to a keyword.

To support the find command, let's also
* change TaskList to be able to find tasks based on a keyword
* allow Parser to parse for find
* allow Ui to display what is found

alvintfl committed:
Merge branch 'branch-A-JavaDoc'

alvintfl committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/tako/Ui.java
#	src/main/java/tako/task/Task.java

alvintfl committed:
Add JavaDocs and unit tests related to the find command

alvintfl committed:
Replace Tako text-based UI with GUI

Follow Level-10 requirements.

alvintfl committed:
Fix bug in storage where the wrong format was being loaded.

alvintfl committed:
DialogBox fxml: display all text with no truncation

For the list command, users will want to see all tasks without the
dialog box truncating the tasks that are further down the list.

alvintfl committed:
Merge branch 'branch-Level-10'

jovitaanderson commented on src/main/java/tako/Parser.java:
```diff
@@ -0,0 +1,148 @@
+    private enum Keyword {
+        BYE, LIST, MARK, TODO, DEADLINE, EVENT, DELETE, FIND;
+
+        private static boolean contains(String s) {
```
Maybe you could change the method name to understand what it should contain? , for example, "hasKeyword" ?

jovitaanderson commented on src/main/java/tako/Parser.java:
```diff
@@ -0,0 +1,148 @@
+            throw new TakoException("The input is invalid.");
+        }
+
+        switch (command) {
```
Good use of switch instead of if/else to prevent deep nesting 🙂

jovitaanderson commented on src/main/java/tako/Tako.java:
```diff
@@ -0,0 +1,58 @@
+        return ui.getResponse();
+    }
+
+    public boolean isExit() {
```
Perhaps you could consider to include javaDoc for isExit() method ? 🤔

jovitaanderson commented on src/main/java/tako/command/AddCommand.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    /**
+     * Returns false as Tako cannot exit after this command.
```
Perhaps you could also consider using @inheritDoc in the javaDoc since it is similar to the Command.java (parent) class

jovitaanderson commented on src/main/java/tako/command/DeleteCommand.java:
```diff
@@ -0,0 +1,53 @@
+        ui.showDelete(task, tasks.getSize());
+    }
+
+    /**
```
Perhaps you could consider using @inheritDoc in the javaDoc here (as well as other children of the command class, exitcommand, findcommand etc.)

jovitaanderson reviewed (commented):
Looks good, I liked how you properly organized your code into it's relevant packages👍🏼 
Just some minor variable naming and javaDocs
LGTM 😄

ou4jairmeiogusrjejiepr commented on src/main/java/tako/Storage.java:
```diff
@@ -0,0 +1,138 @@
+    private String convertToFileFormat(Task task) {
+        String s = task.toString();
+        char taskType = s.charAt(1);
+        int isDone = s.charAt(4) == ' ' ? 0 : 1;
```
I like this, it's a very clean way to check if the task is done!

ou4jairmeiogusrjejiepr commented on src/main/java/tako/Storage.java:
```diff
@@ -0,0 +1,138 @@
+                    splitLine[3], formatter);
+            task = (new Event(splitLine[2], eventDateTime));
+            break;
+        default:
```
I think this can be omitted, since it is empty. If you're leaving it intentionally to improve readability, that's fine too

ou4jairmeiogusrjejiepr commented on src/main/java/tako/Ui.java:
```diff
@@ -0,0 +1,97 @@
+     * Informs the user that a loading error has occurred.
+     */
+    public void showLoadingError() {
+        response = "Tasks failed to load." + "A new task list will be used instead.";
```
I think you can remove the concatenation here

ou4jairmeiogusrjejiepr commented on src/test/java/tako/ParserTest.java:
```diff
@@ -0,0 +1,173 @@
+import tako.command.ListCommand;
+import tako.command.MarkCommand;
+
+public class ParserTest {
```
Great test cases, very comprehensive

ou4jairmeiogusrjejiepr reviewed (commented):
LGTM, very clean code overall!

alvintfl committed:
Add assertions

There are several invariants of some methods that should hold true.

There is nothing to check that these invariants are true.

Let's add assertions to check these invariants.

Assertions can be disabled without modifying the code.

alvintfl committed:
Improve code quality

Some parts of the code are not very readable due to the following
reasons:
* long method
* use of magic numbers

Let's extract out methods from code fragments that are grouped
together and use named constants to explain magic numbers.

This will make the code more readable.

alvintfl committed:
Merge pull request #1 from alvintfl/branch-A-Assertions

Add assertions

alvintfl committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/tako/Parser.java
#	src/main/java/tako/Storage.java

alvintfl committed:
Merge pull request #2 from alvintfl/branch-A-CodeQuality

Branch a code quality

alvintfl committed:
Add more assertions

alvintfl committed:
Fix style issues

alvintfl committed:
Add sort command to sort task list

Users can only see their tasks in the order they were added in.
They may want to see which tasks they want to finish urgently or
prefer a different ordering.

Let's add a sort command to sort the task list.

This allows the user to provide arguments to specify how they
would want to sort the task list.

alvintfl committed:
Add unit test for sort() in TaskList and parsing sort in Parser

There are no tests to verify that the new sort command works and
whether or not it breaks the old features.

Let's add unit test for the sort command.

alvintfl committed:
Merge branch 'branch-BCD-Extension'

alvintfl committed:
Set up Continuous Integration

alvintfl committed:
Remove I/O redirection tests from CI

alvintfl committed:
Fix unit test import statement ordering

alvintfl committed:
Give credit for reused work

alvintfl committed:
Add title to GUI

alvintfl committed:
Add picture of ui

alvintfl committed:
Update README.md

alvintfl committed:
Fix Ui.png where search bar was not displayed properly

alvintfl committed:
Fix header comments
